agent/agent.go

```diff
@@ -0,0 +1,206 @@
+//  Copyright (c) 2018 Cisco and/or its affiliates.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at:
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package agent
+
+import (
+	"os"
+	"os/signal"
+	"syscall"
+
+	"errors"
+
+	"github.com/ligato/cn-infra/core"
+	"github.com/ligato/cn-infra/logging/logrus"
+	"github.com/ligato/cn-infra/utils/once"
+	"github.com/namsral/flag"
+)
+
+// Agent implements startup & shutdown procedures.
+type Agent interface {
+	Run() error
+	Start() error
+	Wait() error
+	After() <-chan struct{}
+	Error() error
+	Stop() error
+	Options() Options
+}
+
+// NewAgent creates a new Agent
+func NewAgent(opts ...Option) Agent {
+	options := newOptions(opts...)
+
+	flag.Parse()
+
+	return &agent{
+		opts: options,
+	}
+}
+
+type agent struct {
+	opts      Options
+	stopCh    chan struct{}
+	startOnce once.ReturnError
+	stopOnce  once.ReturnError
+}
+
+// Options returns the Options the agent was created with
+func (a *agent) Options() Options {
+	return a.opts
+}
+
+// Start starts the agent.  Start will return as soon as the Agent is ready.  The Agent continues
+// running after Start returns.
+func (a *agent) Start() error {
+	return a.startOnce.Do(a.startSignalWrapper)
```

Comment:
I can't seem to see any logic behind this. "_Protecting the code_" this way just encourages incorrect use and adds extra layer of indirection. And more importantly this patter over `Start`/`Stop` prevents `agent` to be stopped and started again. So, this doesn't really make sense to allow multiple calls to start, but prevent to actually start again after stopping correctly.

You could simply do check if `stopCh` is `nil` and return error, since that's a way you are already using in other methods to check if agent was started or not. That is actually similar way how it's done in `os/exec` package in standard libary.

Reply:
The fundamental question comes back to I think the following things:

1)  Do you want to recycle Agents (ie, be able to restart a stopped agent)
2) Are you willing to accept things going wrong if Start or Stop are called more than once?
3) Are you willing to accept things going wrong is Start and Stop are called while one or the other of them is running?
4) Are you willing to accept getting returned errors if start or stop are called more than once?
5) Are you willing to accept things going wrong if there is a race condition between two thread?

My answers would be no to all 5.  If you answer no to all five, then you can either:

a)  Use once.Do, which elegantly does 2-5.
b)  Protect stopCh with a mutex and use it as you suggest
c)  Introduce and mutex protect places to put startErr and stopErr
At which point you've reinvented once.Do, but in a much more confusing way.

Incorrect use is not something you can discourage, especially with potential race conditions.  It happens, and if you are lucky it blows up immediately.  If you are unlucky, it blows up unpredictably at random in the future.

Reply:
1. Depends.. someone might have use case for this, don't see technical limit of why it couldn't be allowed.

2. Yes, I would expect getting an error on second call to `Start` or `Stop`. Which would not hide the incorrect usage from devs and make it immediately obvious what they are doing wrong and make them fix their code instead of keeping pointless multiple calls to `Start`.
  
    This is similar way how they deal with this in **standard library**, calling `Start` for [exec.Cmd](https://godoc.org/os/exec#Cmd) while the process is already running will [return explicit error](https://golang.org/src/os/exec/exec.go#L354) that it was already started.

3. Yes, I expect things going wrong when I use some library *incorrectly*. Calling `Stop` while `Start` is running should not happen in *correct* code, meaning the normal usage dictates that we should call one after another or more specifically accessing agent only from one goroutine.

4. Explained in 2.

5. Definitely yes! I expect things going wrong when there is race condition! Race condition is a bug and should not be allowed to be kept in the code. You should be able to use tools to discover and fix them.
  
These kind of problems you are describing are also possible with `exec.Cmd` and I'm sure with ton of other things too. However they are not trying to "*protect the code*" from this. There aren't any `sync.Once` or mutexes in [(*Cmd).Start](https://golang.org/src/os/exec/exec.go#L338) method, which would allow calls from multiple goroutines. 

Sure, there are use cases for `sync.Once` even in the `os/exec` package, but they are mostly to prevent closing of files, specifically `stdin` pipe which might behave differently for different types of commands and not to allow accessing it from multiple goroutines.

By not "*protecting the code*" from these things you can easily **detect incorrect usage** of the program by running it with `-race` flag. With your pattern this would **not** be possible, because the `sync.Once` would hide this and would not uncover such incorrect code.